engine/src/main/java/org/terasology/rendering/dag/StateManager.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2016 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.rendering.dag;
+
+
+import com.google.api.client.util.Maps;
+import java.util.Collection;
+import java.util.Map;
+import org.terasology.context.Context;
+import org.terasology.rendering.dag.states.State;
+import org.terasology.rendering.dag.states.StateType;
+import org.terasology.rendering.dag.states.StateTypeImpl;
+import org.terasology.rendering.dag.states.StateValue;
+import org.terasology.rendering.dag.states.Wireframe;
+
+/**
+ * TODO: add javadocs
+ */
+public class StateManager {
+
+    private Context context;
+    private Map<String, Map<StateType, StateValue>> stateChanges; // TODO: better name?
+    private Map<StateType, State> availableStates;
+
+    public StateManager() {
+        initialise();
+    }
+
+    public StateManager(Context context) {
+        this.context = context;
+        initialise();
+    }
+
+    private void initialise() {
+        stateChanges = Maps.newHashMap();
+        availableStates = Maps.newHashMap();
+    }
+
+    public static StateManager createDefault(Context context) {
+        StateManager stateManager = new StateManager(context);
+        stateManager.addAvailableStates();
+        return stateManager;
+    }
+
+
+    public void addState(StateType stateType, State state) {
+        availableStates.put(stateType, state);
+    }
+
+    public void findStateChanges(Collection<Node> nodes) {
```

Comment:
The term "find" is a bit limited compared to the goal of this method. This method intends to prevent redundant state changes from occurring and should be renamed accordingly, i.e. "eliminateRedundantStateChanges".

That been said, I doubt it will take this form in the RenderPipelineOptimizer as an instance of that class will work on a list of StateChanges and Nodes, working sequentially and eliminating state changes while it iterates through them. In this method you are working with chunks of state changes (desired state changes vs planned state changes for a given node), making the method unnecessarily complex. 

We might re-introduce some complexity eventually, when we'll have to track changing node parameters and their impact on the pipeline. But at this stage I'd keep it simple.